app/src/main/kotlin/com/wire/android/ui/common/textfield/CodeTextField.kt

```diff
@@ -81,101 +61,94 @@ fun CodeTextField(
+        shape = shape,
+        colors = colors,
+        textStyle = textStyle,
+        state = state,
+        maxHorizontalSpacing = maxHorizontalSpacing,
         horizontalAlignment = horizontalAlignment,
-        modifier = modifier.width(IntrinsicSize.Min),
-    ) {
-        BasicTextField(
-            value = value,
-            onValueChange = {
-                val textDigits = it.text.filter { it.isDigit() } // don't allow characters other than digits to be entered
-                    .let { it.substring(0, min(codeLength, it.length)) } // don't allow more digits than required
-                onValueChange(
-                    CodeFieldValue(
-                        text = TextFieldValue(text = textDigits, selection = TextRange(textDigits.length)),
-                        isFullyFilled = textDigits.length == codeLength
-                    )
-                )
-            },
-            enabled = enabled,
-            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number, autoCorrect = false, imeAction = ImeAction.Done),
-            keyboardActions = KeyboardActions(onDone = { keyboardController?.hide() }),
-            interactionSource = interactionSource,
-            decorationBox = {
-                Row(horizontalArrangement = Arrangement.SpaceBetween) {
-                    repeat(codeLength) { index ->
-                        if (index != 0) Spacer(modifier = Modifier
-                            .weight(1f, fill = false)
-                            .width(maxHorizontalSpacing))
-                        Digit(
-                            char = value.text.getOrNull(index),
-                            shape = shape,
-                            colors = colors,
-                            textStyle = textStyle,
-                            selected = index == value.text.length,
-                            state = state
-                        )
-                    }
-                }
-            })
-        val bottomText = when {
-            state is WireTextFieldState.Error && state.errorText != null -> state.errorText
-            else -> String.EMPTY
-        }
-        AnimatedVisibility(visible = bottomText.isNotEmpty()) {
-            Text(
-                text = bottomText,
-                style = MaterialTheme.wireTypography.label04,
-                color = colors.descriptionColor(state).value,
-                modifier = Modifier
-                    .padding(top = MaterialTheme.wireDimensions.spacing4x)
+        modifier = modifier,
+        innerBasicTextField = { decorator, textFieldModifier ->
+            BasicTextField(
+                state = textState,
+                textStyle = textStyle,
+                enabled = enabled,
+                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number, autoCorrect = false, imeAction = ImeAction.Done),
+                keyboardActions = KeyboardActions(onDone = { keyboardController?.hide() }),
```

Comment:
Can't we use `onImeAction` here as we did for others?

```
onImeAction = { keyboardController?.hide() },
```

Reply:
> Can't we use `onImeAction` here as we did for others?
> 
> ```
> onImeAction = { keyboardController?.hide() },
> ```

Wouldn't it also dismiss the keyboard when the user types on the `Next` IME action (that is, going from one field to the other)?
![image](https://github.com/wireapp/wire-android/assets/9389043/8ecff2fc-11d6-44d6-ac50-e9c4a28dc2b3)

Reply:
Our `CodeTextField` uses `BasicTextField` which takes `keyboardActions: KeyboardActions`.
We use `onImeAction` for `WirePasswordTextField` only, because it uses `BasicSecureTextField` instead of `BasicTextField` and it doesn't allow to use `KeyboardActions`, instead, it takes `onSubmit: ImeActionHandler? = null`.
Actually in `BasicSecureTextField` there is a TODO comment: 
`// TODO(b/297425359) Investigate cleaning up the IME action handling APIs.`
so probably they will introduce a better solution to that soon. 😄